graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java

```diff
@@ -93,48 +93,42 @@ protected HttpResponse executeGraphql(ServiceRequestContext ctx, GraphqlRequest
         } else {
             future = graphQL.executeAsync(input);
         }
-        return HttpResponse.from(future.handle((executionResult, cause) -> {
-            if (cause != null) {
-                // graphQL.executeAsync() returns an error in the executionResult with getErrors().
-                // Use 500 Internal Server Error because this cause might be unexpected.
-                final ExecutionResult error = newExecutionResult(cause);
-                return HttpResponse.ofJson(HttpStatus.INTERNAL_SERVER_ERROR, produceType,
-                                           error.toSpecification());
-            }
-            final List<GraphQLError> errors = executionResult.getErrors();
-            if (!errors.isEmpty() && errors.stream().anyMatch(ValidationError.class::isInstance)) {
-                // The server SHOULD deny execution with a status code of 400 Bad Request for
-                // invalidate documentation.
-                return HttpResponse.ofJson(HttpStatus.BAD_REQUEST, produceType,
-                                           executionResult.toSpecification());
-            }
-            return toHttpResponse(executionResult, produceType);
-        }));
+        return HttpResponse.from(
+                future.handle((executionResult, cause) -> {
+                    if (executionResult.getData() instanceof Publisher) {
+                        logger.warn("executionResult.getData() returns a {} that is not supported yet.",
+                                    executionResult.getData().toString());
+
+                        return HttpResponse.ofJson(HttpStatus.NOT_IMPLEMENTED,
+                                                   produceType,
+                                                   toSpecification("WebSocket is not implemented"));
+                    }
+
+                    if (executionResult.getErrors().isEmpty() && cause == null) {
+                        return HttpResponse.ofJson(produceType, executionResult.toSpecification());
+                    }
+
+                    return errorHandler.handle(ctx, input, executionResult, cause);
+                }));
     }
 
-    private static HttpResponse toHttpResponse(ExecutionResult executionResult, MediaType produceType) {
-        // TODO: When WebSocket is implemented, it should be removed.
-        if (executionResult.getData() instanceof Publisher) {
-            logger.warn("executionResult.getData() returns a {} that is not supported yet.",
-                        executionResult.getData().toString());
-            final ExecutionResult error =
-                    newExecutionResult(new UnsupportedOperationException("WebSocket is not implemented"));
-            return HttpResponse.ofJson(HttpStatus.NOT_IMPLEMENTED, produceType, error.toSpecification());
-        }
-        return HttpResponse.ofJson(produceType, executionResult.toSpecification());
-    }
+    private static Map<String, Object> toSpecification(String message) {
```

Comment:
How about making this package-private and sharing this with `DefaultGraphqlErrorHandler`?